toolchain/check/eval.cpp

```diff
@@ -943,23 +932,155 @@ static auto PerformBuiltinBinaryIntOp(Context& context, SemIRLoc loc,
+      if (!width ||
+          *width > IntStore::MaxIntWidth - lhs_val.getSignificantBits()) {
+        CARBON_DIAGNOSTIC(CompileTimeUnsizedShiftOutOfRange, Error,
+                          "shift distance of {0} would result in an "
+                          "integer whose width is greater than the "
+                          "maximum supported width of {1}",
+                          TypedInt, int);
+        context.emitter().Emit(loc, CompileTimeUnsizedShiftOutOfRange,
+                               {.type = rhs.type_id, .value = rhs_orig_val},
+                               IntStore::MaxIntWidth);
+        return SemIR::ErrorInst::SingletonConstantId;
+      }
+      lhs_val = lhs_val.sext(
+          IntStore::CanonicalBitWidth(lhs_val.getSignificantBits() + *width));
+    }
+
+    result_val =
+        lhs_val.shl(rhs_orig_val.getLimitedValue(lhs_val.getBitWidth()));
+  } else if (lhs_is_signed) {
+    result_val =
+        lhs_val.ashr(rhs_orig_val.getLimitedValue(lhs_val.getBitWidth()));
+  } else {
+    CARBON_CHECK(lhs_bit_width_id.is_valid(), "Logical shift on unsized int");
+    result_val =
+        lhs_val.lshr(rhs_orig_val.getLimitedValue(lhs_val.getBitWidth()));
+  }
+  return MakeIntResult(context, lhs.type_id, lhs_is_signed,
+                       std::move(result_val));
+}
+
+// Performs a homogeneous builtin binary integer -> integer operation.
+static auto PerformBuiltinBinaryIntOp(Context& context, SemIRLoc loc,
+                                      SemIR::BuiltinFunctionKind builtin_kind,
+                                      SemIR::InstId lhs_id,
+                                      SemIR::InstId rhs_id)
+    -> SemIR::ConstantId {
+  auto lhs = context.insts().GetAs<SemIR::IntValue>(lhs_id);
+  auto rhs = context.insts().GetAs<SemIR::IntValue>(rhs_id);
+
+  CARBON_CHECK(rhs.type_id == lhs.type_id, "Heterogeneous builtin integer op!");
+  auto type_id = lhs.type_id;
+  auto [is_signed, bit_width_id] =
+      context.sem_ir().types().GetIntTypeInfo(type_id);
+  auto [lhs_val, rhs_val] =
+      GetIntsAtSuitableWidth(context, lhs.int_id, rhs.int_id, bit_width_id);
+
+  // Check for division by zero.
+  switch (builtin_kind) {
+    case SemIR::BuiltinFunctionKind::IntSDiv:
+    case SemIR::BuiltinFunctionKind::IntSMod:
+    case SemIR::BuiltinFunctionKind::IntUDiv:
+    case SemIR::BuiltinFunctionKind::IntUMod:
+      if (rhs_val.isZero()) {
+        DiagnoseDivisionByZero(context, loc);
+        return SemIR::ErrorInst::SingletonConstantId;
+      }
+      break;
+    default:
+      break;
+  }
+
+  BinaryIntOpResult result =
+      ComputeBinaryIntOpResult(builtin_kind, lhs_val, rhs_val);
+
+  if (result.overflow && !bit_width_id.is_valid()) {
+    // Retry with a larger bit width. Most operations can only overflow by one
+    // bit, but signed n-bit multiplication can overflow to 2n-1 bits.
+    int new_width =
+        builtin_kind == SemIR::BuiltinFunctionKind::IntSMul
```

Comment:
Does this not seem to imply that IntUMul maybe have introduced wrapping on the first `ComputeBinaryIntOpResult()` attempt, and wouldn't if we'd increased its bitwidth? Is that desirable?

Reply:
Unsigned operations aren't meaningful on an unsized integer type, so that's not even possible here. I've extended the comment to explain and added a `CHECK`.